### ISSUE TITLE
NONE window_decorations does not work on GNOME 40 Wayland

### ISSUE THREAD

**Issue by almereyda** [wez/wezterm#1077]
### What Operating System(s) are you seeing this problem on?

Linux Wayland

### WezTerm version

20210824-085333-d38ba132

### Did you try the latest nightly build to see if the issue is better (or worse!) than your current version?

Yes, and I updated the version box above to show the version of the nightly that I tried

### Describe the bug

When disabling window decorations with Wayland enabled, the title bar is not hidden and displayed in an odd location.

![Bildschirmfoto von 2021-08-24 22-16-49](https://user-images.githubusercontent.com/1645308/130684003-3512647d-06df-4488-9a38-a5d6843dc637.png)

This is because the actual terminal pane will only resize to the full window size when been having been hovered.

![Bildschirmfoto von 2021-08-24 21-56-44](https://user-images.githubusercontent.com/1645308/130683209-c4b550a8-09d1-449a-885c-e7dedbc17835.png)

The post-creation resizing only works if debug mode is not active.
It will remain available as a micro terminal until touched, while the title bar will not be part of the application window in a focussed screen shot (Alt + Print):

![Bildschirmfoto von 2021-08-24 22-10-23](https://user-images.githubusercontent.com/1645308/130683181-bb508f84-c1d1-4b2b-bc96-12f2cad54be0.png)

### To Reproduce

On Fedora 34 with GNOME 40 on Wayland, run `wezterm` with the configuration below and experience the described glitches.

### Configuration

```
  window_decorations = "NONE",
  enable_wayland = true,
```

### Expected Behavior

WezTerm displays itself similarly to what we see when using the configuration

```
  window_decorations = "NONE",
  dpi = 192.0,
```

through XWayland.

![Bildschirmfoto von 2021-08-24 22-15-23](https://user-images.githubusercontent.com/1645308/130683668-b9eb023b-b1a0-4308-97d2-d439e487f353.png)

### Logs

```
 2021-08-24T20:00:13.834Z INFO  wezterm_mux_server_impl::local > setting up /run/user/1000/wezterm/gui-sock-127366
 2021-08-24T20:00:14.094Z INFO  wezterm_gui::termwindow        > OpenGL initialized! Mesa Intel(R) UHD Graphics 620 (KBL GT2) 4.6 (Compatibility Profile) Mesa 21.1.7 is_context_loss_possible=false wezterm version: 20210824-085333-d38ba132
 ```

**Comment by wez** [comment]
Note that https://wezfurlong.org/wezterm/config/lua/config/window_decorations.html states that `window_decorations` is not supported on Wayland.

I think the other things you describe overlap with https://github.com/wez/wezterm/issues/1020

As discussed in that other issue, I can't run Wayland on my nvidia hardware so progress on annoying issues like these is very slow: I have empathy for them, but trying to improve them is about 10x harder and more time consuming than me spending my spare time on other things.

I'd love to accept contributions from others that improve this!

**Comment by almereyda** [comment]
Thank you @wez for the detailed explanation. This is exactly why I was asking.

Please allow some time to pass, before a voluntary contribution can come into existence here. Yet this sounds like a fun thing to explore, when rainy and dark autumn days arrive.

**Comment by EpocSquadron** [comment]
@wez it's also worth noting that there is a [separate library](https://gitlab.gnome.org/jadahl/libdecor/-/tree/master) that popped up recently to provide client side decorations under wayland. I don't see any bindings for rust yet, and i don't know if smithay would have to pull it in or if wezterm could use it on the side, but I figure it's worth a peek.

**Comment by wez** [comment]
I saw libdecor, but it seems like it is a distinctly limited subset of the frame drawing we inherited from SCTK.  I don't see there being any value in targeting libdecor in wezterm.

**Comment by unrelentingtech** [comment]
> the actual terminal pane will only resize to the full window size when been having been hovered.

Oh. This is a scaling issue I've just encountered as #1111 and fixed as #1112. Could not find this report because it mentioned nothing about HiDPI.

> I can't run Wayland on my nvidia hardware

You don't have to run directly on hardware, you can run most compositors nested in a window (on X11). At least wlroots ones (Wayfire, sway, …) and weston.

> used to provide an implementation of client-side decorations but recently deprecated and removed it

Actually they still have `FallbackFrame`, which I think is what winit still uses.

> I don't see there being any value in targeting libdecor in wezterm.

Well, it seems like the eventual goal of libdecor is to have desktop environments provide libdecor plugins so any apps using libdecor could achieve that DE-native look…?

---

just stumbled upon these lines https://github.com/rust-windowing/winit/blob/1b3b82a3c1369c5248a2e6a251230ba6c615e918/src/platform_impl/linux/wayland/window/mod.rs#L226-L230 I wonder if they could be relevant to anything

**Comment by wez** [comment]
>> I can't run Wayland on my nvidia hardware
> You don't have to run directly on hardware, you can run most compositors nested in a window (on X11). At least wlroots ones (Wayfire, sway, …) and weston.

I've tried those, but running that way is not the same as running the compositor directly; the behavior is subtly different and each of them is differently magical :-/

> Actually they still have FallbackFrame, which I think is what winit still uses.

It is, but it doesn't have the ability to render text in the title bar, so we have those bits and connect them to our font rasterizer.
It's still far from perfect!

> Well, it seems like the eventual goal of libdecor is to have desktop environments provide libdecor plugins so any apps using libdecor could achieve that DE-native look…?

At that point it will become more interesting, but it still seems pretty bare bones today!
